### ISSUE TITLE
Please investigate and improve back propagation

### ISSUE THREAD

**Issue by nleck** [stSoftwareAU/NEAT-TS#1869]
Back propagation isn't improving our large complex creature score as much as our other evolutionary techniques. I think we have some issues with complex creatures.

Can we test cases to check corner cases and hopefully shed some light on why back propagation for large complex creatures with all the squash functions and combinations there of.

I'm after ways of improving our production creature score whose size is "Observations: 2,356, Neurons: 1,089 & Synapses: 18,873"

**Comment by stservice** [comment]
## Planning Summary

After thorough analysis of the back propagation implementation, I've identified several likely causes for reduced effectiveness on large complex creatures and created 6 sub-issues to investigate and address them.

### Key Findings

**Root cause analysis** of why backprop underperforms for the production creature (2,356 observations, 1,089 neurons, 18,873 synapses):

1. **Gradient magnification** — In `TopologicalBackpropagation.ts:92`, error signals from all downstream paths are summed without normalisation. The `targetDeltaCount` is tracked (line 276) but never used. Neurons with 50+ downstream connections receive 50× the gradient of neurons with 1 connection, causing oversized weight updates.

2. **Dead gradient paths via near-zero weights** — When synapse weight ≈ 0, the effective weight division (line 259-261) creates extreme target activations (e.g., 100,000+) that exceed the squash range and get silently dropped (lines 268-269). In large networks, many synapses have near-zero weights, blocking gradient flow through significant portions of the network.

3. **Aggregate function dead paths** — MAXIMUM/MINIMUM neurons use winner-take-all gradient flow — only the winning connection receives error. All other connections never learn via backprop.

4. **Mixed squash function gradient scale mismatches** — The production creature uses 30+ squash types. Chains of different squash functions create inconsistent gradient scales, and there are no tests covering these mixed chains.

### Sub-Issues Created

| # | Issue | Type | Dependencies |
|---|-------|------|-------------|
| 1 | #1870 — Test: Mixed squash function chain backprop corner cases | Diagnostic | None |
| 2 | #1871 — Test: Gradient accumulation scaling in highly-connected neurons | Diagnostic | None |
| 3 | #1872 — Enhancement: Normalise gradient accumulation for high fan-out neurons | Fix | #1870, #1871 |
| 4 | #1873 — Enhancement: Improve near-zero weight error propagation stability | Fix | #1870, #1871 |
| 5 | #1874 — Enhancement: Improve MAXIMUM/MINIMUM aggregate function gradient flow | Fix | #1870, #1871 |
| 6 | #1875 — Test: Production-scale creature backprop convergence validation | Validation | #1872, #1873, #1874 |

### Implementation Order

```
Phase 1 (parallel):  #1870 + #1871  — Diagnostic tests to verify hypotheses
Phase 2 (parallel):  #1872 + #1873 + #1874  — Targeted improvements
Phase 3:             #1875  — End-to-end validation at production scale
```

### Assumptions

- The `targetDeltaCount` variable being tracked but unused (TopologicalBackpropagation.ts:276) is a genuine gap, not a deliberate design choice
- Default behaviour should remain backwards-compatible — new normalisation should be opt-in via configuration
- The production creature likely uses all major squash categories including aggregate functions (IF, MAXIMUM, MINIMUM)
- Convergence improvements can be measured within unit test iteration counts (not requiring hours of training)

---

This planning issue will be automatically closed. Reopen it if you'd like to adjust the plan.

**Comment by stservice** [comment]
Planning complete. Sub-issues were detected but specific URLs could not be extracted from the output.
---

🤖 Processed by: stservice
